src/server/tiered_storage.cc

```diff
@@ -241,19 +260,31 @@ util::fb2::Future<string> TieredStorage::Read(DbIndex dbid, string_view key,
                                               const PrimeValue& value) {
   DCHECK(value.IsExternal());
   util::fb2::Future<string> future;
-  auto cb = [future](string* value) mutable {
-    future.Resolve(*value);
-    return false;
+
+  // The raw_val passed to cb might need decoding based on the encoding mask of the "value" object.
+  // We save the mask in decoder and use it to decode the final string that Read should resolve.
+  PrimeValue decoder;
+  decoder.ImportExternal(value);
+
+  auto cb = [future, decoder = std::move(decoder)](bool is_raw, const string* raw_val) mutable {
+    future.Resolve(DecodeString(is_raw, *raw_val, std::move(decoder)));
+    return false;  // was not modified
   };
```

Comment:
weird that we need this decoder, if it had a proper static interface it would be easier

Reply:
no, because we need the encoding mask of the object. Per your suggestion we now do not store it on disk so we need to keep it in memory to be able to decode a blob. decoder calls `ImportExternal(value);` that copies the encoding mask.

Reply:
We could expose the internals and fetch the mask but I thought this interface is more "elegant".

Reply:
I'm a little confused... Inside ::materialize you never use the mask, only when calling SetMeta(), but there you already are beyound the conditional